src/Servers/Kestrel/Transport.Quic/src/Internal/QuicStreamContext.cs

```diff
@@ -0,0 +1,337 @@
+                if (_connectionId == null)
+                {
+                    _connectionId = $"{_connection.ConnectionId}:{StreamId}";
+                }
+                return _connectionId;
+            }
+            set
+            {
+                _connectionId = value;
+            }
+        }
+
+        private async Task StartAsync()
+        {
+            try
+            {
+                // Spawn send and receive logic
+                var receiveTask = DoReceive();
+                var sendTask = DoSend();
+
+                // Now wait for both to complete
+                await receiveTask;
+                await sendTask;
+
+            }
+            catch (Exception ex)
+            {
+                _log.LogError(0, ex, $"Unexpected exception in {nameof(QuicStreamContext)}.{nameof(StartAsync)}.");
+            }
+        }
+
+        private async Task DoReceive()
+        {
+            Exception error = null;
+
+            try
+            {
+                await ProcessReceives();
+            }
+            catch (QuicException ex)
+            {
+                // This could be ignored if _shutdownReason is already set.
+                error = new ConnectionResetException(ex.Message, ex);
+            }
+            catch (Exception ex)
+            {
+                // This is unexpected.
+                error = ex;
+                _log.StreamError(ConnectionId, error);
+            }
+            finally
+            {
+                // If Shutdown() has already bee called, assume that was the reason ProcessReceives() exited.
+                Input.Complete(_shutdownReason ?? error);
+
+                FireStreamClosed();
+
+                await _waitForConnectionClosedTcs.Task;
+            }
+        }
+
+        private async Task ProcessReceives()
+        {
+            var input = Input;
+            while (true)
+            {
+                var buffer = Input.GetMemory(MinAllocBufferSize);
+                var bytesReceived = await _stream.ReadAsync(buffer);
+
+                if (bytesReceived == 0)
+                {
+                    // Read completed.
+                    break;
+                }
+
+                input.Advance(bytesReceived);
+
+                var flushTask = input.FlushAsync();
+
+                var paused = !flushTask.IsCompleted;
+
+                if (paused)
+                {
+                    _log.StreamPause(ConnectionId);
+                }
+
+                var result = await flushTask;
+
+                if (paused)
+                {
+                    _log.StreamResume(ConnectionId);
+                }
+
+                if (result.IsCompleted || result.IsCanceled)
+                {
+                    // Pipe consumer is shut down, do we stop writing
+                    break;
+                }
+            }
+        }
+
+        private void FireStreamClosed()
+        {
+            // Guard against scheduling this multiple times
```

Comment:
This is only called from DoReceive, right?

Reply:
Yeah, this is the same as SocketConnection. Maybe it should be fixed there as well.